### PR TITLE
add JavaScript browser-based compat tests

### PR DESCRIPTION
**Pull request by anurse**
Note: nothing runs them in the build at the moment, but they all pass
when launched in the browser. Next phase will be automated command-line
based running. @moozzyk is also working on command-line-based browser tests so I'm going to coordinate with that.

Part WhoKnows of ∞ for #206 

/cc @BrennanConroy @moozzyk

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -32,4 +32,5 @@ project.lock.json
 runtimes/
 .build/
 .testPublish/
-launchSettings.json
\ No newline at end of file
+launchSettings.json
+node_modules/
```
Comment (moozzyk) [inline .gitignore]: It appears that a convention is to install in bin: https://github.com/aspnet/KoreBuild/blob/12824ac462ec161ce4790decee4e5947344bc61b/build/shade/_bower.shade#L3

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/wwwroot/default.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  <link rel="stylesheet" href="lib/jasmine.css" />
+</head>
+<body>
+  <script src="lib/jquery.js"></script>
+  <script src="lib/jquery.signalR.js"></script>
+  <script src="test/hubs/hubs"></script>
+  <script src="lib/jasmine.js"></script>
+  <script src="lib/jasmine-html.js"></script>
+  <script src="lib/boot.js"></script>
+  <script src="js/tests.js"></script>
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/wwwroot/default.html]: can you create this in a compatTests folder? Because we are going to have a on server and we will probably have many functional suites we would put functional tests into a functionalTests folder (or similar)

Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/wwwroot/default.html]: Yeah, I created this before we had decided on one server. Will do. I do think that we should have separate HTML pages (for now at least)

File: .gitignore (hunk 1 above)
Comment (anurse) [inline .gitignore]: That line is the building the path to the `bower` executable inside the `node_modules` folder. Not sure what you're suggesting :)

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/wwwroot/default.html (hunk 2 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/wwwroot/default.html]: +1 for separate pages. For now we will have just 2 - one for compat and one for functional tests

File: .gitignore (hunk 1 above)
Comment (moozzyk) [inline .gitignore]: Never mind, I think I figured this out.

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/Startup.cs [hunk 3]
```diff
@@ -26,12 +26,7 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
             app.UseWebSockets();
             app.UseSignalR<TestConnection>("/test/raw");
             app.UseSignalR("/test/hubs");
-
-            var data = Encoding.UTF8.GetBytes("Server online");
-            app.Use(async (context, next) =>
-            {
-                await context.Response.Body.WriteAsync(data, 0, data.Length);
-            });
+            app.UseFileServer();
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/Startup.cs]: Isn't StaticFiles enough?

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/foreverFrame.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿describe('foreverFrame transport', () => {
+    it('should fail to initialize connection', done => {
+        let connection = $.connection(HUBS_URL);
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/foreverFrame.ts]: Should this be `CONNECTION_URL`? (otherwise I cannot spot any difference between these tests)

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/typings.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "dependencies": {},
+  "globalDependencies": {
+    "jasmine": "registry:dt/jasmine#2.2.0+20160621224255",
+    "jquery": "registry:dt/jquery#1.10.0+20160704162008",
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/typings.json]: how are these related to actual version we use. e.g. we are using jquery 3.1.0 but these type definitions does not seem to be for 3.x as per: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/jquery/jquery.d.ts#L1
(signalr probably does not matter since there is no API changes in 2.2.1 as far as I am aware)

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/hubConnectionTests.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        };
+    }
+
+    eachTransport(transport => {
+        describe(`over the ${transport.name} transport`, () => {
+            it('can connect to server', done => {
+                var client = createChatHubConnection(transport);
+                return client.start().then(() => {
+                    expect(client.connection.state).toEqual($.signalR.connectionState.connected);
+                }).fail(fail).always(() => {
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/hubConnectionTests.ts]: what is the purpose of `fail` here?

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/Startup.cs (hunk 3 above)
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/Startup.cs]: Yeah, I generally like having DefaultDocument around when I have static files. Once we unify the servers, I expect there will be a default.html that allows people to Ctrl-F5 and choose a test suite to run.

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/typings.json (hunk 5 above)
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/typings.json]: I'm just going to find the right typings and check them in. It's all JavaScript and we don't plan on updating these packages so npm/typings/gulp is getting a bit overkill ;)

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/hubConnectionTests.ts (hunk 6 above)
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/hubConnectionTests.ts]: If the Promise fails, Jasmine doesn't fail the test (it seems). This wires up a handler to the Promise failure that calls Jasmine's own fail method to ensure the test is marked as failing.

Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/hubConnectionTests.ts]: I might look at installing a replacement `it` that does the right thing here though. When the callback returns a promise, jasmine _should_ be capturing failure from that promise as a failing tests. I may also be doing something wrong here.

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            send: (type, sourceOrDest, value) => connection.send(JSON.stringify({
+                type: type,
+                sourceOrDest: sourceOrDest,
+                value: value
+            }))
+        };
+    }
+
+    eachTransport(transport => {
+        describe(`over the ${transport.name} transport`, () => {
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: are backticks correct?

Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: Yes, they are interpolated strings in TypeScript/ES6 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: Thanks. I did not know.

File: test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        connection.logging = true;
+
+        connection.received(data => {
+            deferred.resolve(data as TestConnectionMessage);
+        });
+
+        return {
+            connection: connection,
+            message: deferred.promise(),
+            start: () => connection.start({ transport: transport.name }),
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: Today I found out that phantomjs does not support many ES6 feature like `let`, arrow functions and Promises. We will probably need to transpile to ES5.

Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: Well, this is typescript, so we can just have it compile down to ES5

Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.CompatTests.Server/jstests/connectionTests.ts]: Yes, this is what I am suggesting.

**Comment by moozzyk** [comment]
@anurse FYI: https://github.com/aspnet/SignalR-Server/tree/pawelka/anurse-js-compat
still some issues to fix:
- removed typings for now so there are errors (ignored) when compiling ts
- tests are flakey
- when all compat tests are enabled js test never finish - maybe releated to flakiness (when running directly I saw some weird timeouts and the process was killed before tests finished running)

**Comment by anurse** [comment]
Not relevant any more, right @moozzyk ?

**Comment by moozzyk** [comment]
Yes. This was rebased and merged from a different branch.
